benchmarks/float8/float8_inference_roofline.py

```diff
@@ -0,0 +1,298 @@
+# LICENSE file in the root directory of this source tree.
+
+"""
+This is a script to estimate the benefit from converting a `torch.nn.Linear`
+layer to float8 given a single saturated GPU, by estimating the difference
+in e2e GPU kernel time between:
+1. bf16 gemms in fwd and
+2. float8 gemms in fwd and float8 overhead
+
+The gemm times are estimated either from direct measurements via benchmarks,
+or with a roofline estimation based on TOPS and peak compute bandwidth of an
+NVIDIA H100 or B200.
+
+The float8 overhead times are estimated by counting memory reads and writes
+based on the specified float8 scaling, and estimating that we can achieve
+a certain % of machine peak memory bandwidth when performing these reads and writes.
+"""
+
+import copy
+from typing import Optional
+
+import fire
+import pandas as pd
+import sympy
+import torch
+import torch.nn as nn
+import tqdm
+from torch.profiler import ProfilerActivity, profile
+from utils import (
+    get_gpu_kernel_gemm_time_s,
+    get_name_to_shapes_iter,
+    profiler_output_to_filtered_time_by_kernel_name,
+)
+
+import torchao
+from torchao.quantization.quant_api import (
+    Float8DynamicActivationFloat8WeightConfig,
+    PerRow,
+    quantize_,
+)
+from torchao.quantization.quantize_.common import KernelPreference
+from torchao.testing.training.roofline_utils import (
+    get_inference_float8_mem_sympy,
+    get_inference_gemm_time_sympy,
+)
+from torchao.utils import is_MI300
+
+
+@torch.no_grad()
+def get_gpu_kernel_time(m, x):
+    # warm up
+    for _ in range(2):
+        __ = m(x)
+
+    # capture a profiling run
+    activities = [ProfilerActivity.CPU, ProfilerActivity.CUDA]
+    n_iter = 5
+    with profile(activities=activities) as prof:
+        for _ in range(n_iter):
+            __ = m(x)
+            torch.cuda.synchronize()
+    # get the gpu kernel time and aggregate it
+    num_leaf_tensors = 1 + len(list(m.parameters()))
+    ref_times = profiler_output_to_filtered_time_by_kernel_name(
+        prof, n_iter, num_leaf_tensors
+    )
+    total_time_s = sum(v for v in ref_times.values()) / 1e6 / n_iter
+    return total_time_s
+
+
+def get_gemm_times(
+    M: int,
+    K: int,
+    N: int,
+    fast_accum: bool,
+    float8_recipe_name: Optional[str],
+):
+    device = torch.device("cuda")
+
+    # bf16 time
+    x_bf16 = torch.randn(M, K, dtype=torch.bfloat16, device=device)
+    # w_bf16 = torch.randn(K, N, dtype=torch.bfloat16, device=device).t().contiguous().t()
+    w_bf16 = torch.randn(K, N, dtype=torch.bfloat16, device=device)
+
+    bf16_time_s = get_gpu_kernel_gemm_time_s(torch.mm, x_bf16, w_bf16)
+
+    e4m3_dtype = torch.float8_e4m3fn
+    if torch.version.hip and torch.cuda.is_available() and is_MI300():
+        e4m3_dtype = torch.float8_e4m3fnuz
+    d1, d2, d3 = e4m3_dtype, e4m3_dtype, torch.bfloat16
+    A = torch.randint(0, 255, (M, K), device=device, dtype=torch.uint8).view(d1)
```

Comment:
nit: I stil lthink that randn distributed tenors in float32 casted is more representative, but have not measured if any meaningful impact on perf